lib/grape/formatter/json_api_pagination.rb

```diff
@@ -0,0 +1,209 @@
+# frozen_string_literal: true
+
+module Grape
+  module Formatter
+    # This an extended and refactored version of the JSONAPI resources formatter
+    # that adds pagination links if there is a wrapper e.g. Kaminari that provides
+    # the appropriate pagination methods.
+    class JsonApiPagination
+      class << self
+        def call(resource, env)
+          new.serialize_resource(resource, env) || Grape::Formatter::Json.call(resource, env)
+        end
+      end
+
+      attr_reader :env, :resource, :endpoint
+
+      def serialize_resource(resource, env)
+        @env = env
+        @resource = resource
+        @endpoint = env['api.endpoint']
+
+        if resource_class.nil?
+          return blank_resource_response(jsonapi_options, resource)
+        end
+
+        build_main_json_output_data
+        add_pagination_links if resource_supports_pagination?
+
+        json_output.to_json
+      end
+
+      private
+
+      # Ensure sort order is maintained, serialize_to_hash can reorder objects if
+      # objects of the array are of different types (polymorphic cases)
+      def build_main_json_output_data
+        if sorted_primary_ids && (data = json_output['data']).present?
+          json_output['data'] = data.sort_by { |d| sorted_primary_ids.index(d['id']) }
+        end
+      end
+
+      def resource_supports_pagination?
+        resource.respond_to?(:current_page) && resource.respond_to?(:total_pages)
+      end
+
+      def add_pagination_links
+        json_output['links'] ||= {}
+
+        add_link_to_json_output('self', original_uri)
+
+        last_page_uri = build_query_for_different_page(resource.total_pages)
+        add_link_to_json_output('last', last_page_uri)
+        first_page_uri = build_query_for_different_page(1)
+        add_link_to_json_output('first', first_page_uri)
+
+        add_previous_page_link if current_page_is_not_the_first_page
+        add_next_page_link if current_page_is_not_the_last_page
+      end
+
+      def add_next_page_link
+        next_or_last_page_number = [resource.total_pages, resource.current_page.succ].min
+        next_page_uri = build_query_for_different_page(next_or_last_page_number)
+        add_link_to_json_output('next', next_page_uri)
+      end
+
+      def add_previous_page_link
+        first_or_previous_page_number = [1, resource.current_page.pred].max
+        prev_page_uri = build_query_for_different_page(first_or_previous_page_number)
+        add_link_to_json_output('prev', prev_page_uri)
+      end
+
+      def old_build_query_for_different_page(query)
+        original_uri.dup.tap { |uri| uri.query = build_query(query) }
+      end
```

Comment:
Can't see anywhere this method is called.

Reply:
Oops. This was part of my therapeutic refactoring which I forgot to delete. http://www.kytrinyx.com/talks/therapeutic-refactoring/

Reply:
Gone